activerecord/lib/active_record/autosave_association.rb

```diff
@@ -383,6 +383,7 @@ def association_valid?(association, record)
         end
 
         if association.options[:autosave]
+          return if record == self
```

Comment:
```suggestion
          return if equal?(record)
```
I think would be more strict? I'm a bit worried about the consequence of using `==` for Active Record.